soroban-env-host/src/native_contract/token/metadata.rs

```diff
@@ -13,31 +16,41 @@ const METADATA_KEY: &str = "METADATA";
 #[contracttype]
 pub struct TokenMetadata {
     pub decimal: u32,
-    pub name: Bytes,
-    pub symbol: Bytes,
+    pub name: String,
+    pub symbol: String,
 }
 
 pub const DECIMAL: u32 = 7;
 
 pub fn set_metadata(e: &Host) -> Result<(), HostError> {
-    let name_and_symbol: (Bytes, Bytes) = match read_asset_info(e)? {
+    let name_and_symbol: (String, String) = match read_asset_info(e)? {
         AssetInfo::Native => {
-            let n = Bytes::try_from_val(e, &e.bytes_new_from_slice(b"native")?)?;
+            let n = String::try_from_val(e, &e.string_new_from_slice("native")?)?;
             (n.clone(), n)
         }
         AssetInfo::AlphaNum4(asset) => {
-            let symbol: Bytes = asset.asset_code.into();
-            let mut name = symbol.clone();
-            name.push(b':')?;
-            name.append(asset.issuer.into())?;
-            (name, symbol)
+            let symbol: String = asset.asset_code;
+            let mut name = symbol.copy_to_rust_string(e)?;
+            name.push(':');
+            let k = ed25519::PublicKey::from_payload(asset.issuer.to_array()?.as_slice())
+                .map_err(|_| ConversionError)?;
+            name.push_str(k.to_string().as_str());
+            (
+                String::try_from_val(e, &e.string_new_from_slice(name.as_str())?)?,
+                symbol,
+            )
         }
         AssetInfo::AlphaNum12(asset) => {
-            let symbol: Bytes = asset.asset_code.into();
-            let mut name = symbol.clone();
-            name.push(b':')?;
-            name.append(asset.issuer.into())?;
-            (name, symbol)
+            let symbol: String = asset.asset_code;
+            let mut name = symbol.copy_to_rust_string(e)?;
+            name.push(':');
+            let k = ed25519::PublicKey::from_payload(asset.issuer.to_array()?.as_slice())
+                .map_err(|_| ConversionError)?;
```

Comment:
`from_payload` has a failure case, but you don't need a failure case for creating the strkey. You can do `ed25519::PublicKey(asset.issuer.to_array()?)`. The composed contents are pub.

Reply:
I also have a different concern about that: `from_payload` operation is not metered.

Reply:
Updated